File: Lesson 7, Stacks and Queues/Lesson 7, Stacks and Queues, StoneWall.c
#include <stdlib.h>     /* qsort */

int compare (const void * a, const void * b)
{
  return ( *(int*)a - *(int*)b );
}

int solution(int H[], int N)
{
    if (N == 1)
        return 1;
        
    int blocks = 0, j = 0, stack[100000];
    stack[0] = H[0];
    for (int i = 1; i < N; i++) {
        if (H[i] < stack[j]) {
            blocks++;
            stack[j] = H[i];
            if (j - 1 >= 0 && stack[j-1] == stack[j])
                j--;
            for (int i = 0; i < j; i++) {
                if (stack[i] != 0 && stack[i] > stack[j]) {
                    blocks++;
                    stack[i] = 0;
                }
            }
        }
        else if (H[i] > stack[j])
            stack[++j] = H[i];
    }
    /*qsort (stack, j+1, sizeof(int), compare);
    int k = 0;
    for (int i = 0; i < j; i++)
        if (stack[i] != stack[i+1])
            k++;
    */
    return blocks + 1;
}

File: Lesson 7, Stacks and Queues/Lesson 7, Stacks and Queues, StoneWall.py
def solution(H):
    blocks, stack = 0, []
    while len(H) > 0:
        stack.append(H[-1])
        for i in range(len(H)-2, -1, -1):
            
            if H[len(H)-1] < H[i]:
                blocks += 1
                del H[i]
    return blocks + len(stack)

"""
def solution(H):
    blocks, stack = 0, [H[0]]
    for i in range(1, len(H)):
        if H[i] < stack[-1]:
            stack.append(H[i])
            j = 0
            while j < len(stack)-1:
                if stack[j] > stack[-1]:
                    blocks += 1
                    del stack[j]
                elif stack[j] == stack[-1]:
                    del stack[j]
                else:
                    j += 1
        elif H[i] > stack[-1]:
            stack.append(H[i])
    return blocks + len(stack)
"""